File: frontend/hex_renderer.py
"""frontend/hex_renderer.py — Converts a GameState snapshot to an SVG string.

The returned string is a complete <svg>…</svg> element intended for injection
into a NiceGUI ui.html() element.  No external assets required.
"""
from __future__ import annotations
import math

from engine.game_state import GameState
from engine.hex_grid import axial_to_pixel, flat_hex_corners
import config


# ── Color tables ──────────────────────────────────────────────────────────────

_TILE_FILL: dict[int, str] = {
    0: config.TILE_NONE_COLOR,
    1: config.PLAYER_TILE_COLORS[1],
    2: config.PLAYER_TILE_COLORS[2],
}

_TILE_STROKE: dict[int, str] = {
    0: config.TILE_STROKE_COLOR,
    1: "#8b2e06",
    2: "#065066",
}

_BOT_FILL   = config.PLAYER_BOT_COLORS
_BOT_BRIGHT = config.PLAYER_BRIGHT_COLORS


def render_hex_grid(state: GameState, hex_size: float = 26.0) -> str:
    """Return a full <svg> string representing the current game state."""

    padding = hex_size * 1.8  # breathing room around the grid

    # ── Compute pixel centers and canvas bounds ───────────────────────────
    centers: dict = {}
    for h in state.grid:
        centers[h] = axial_to_pixel(h.q, h.r, hex_size)

    xs = [cx for cx, _ in centers.values()]
    ys = [cy for _, cy in centers.values()]
    half_w = hex_size
    half_h = hex_size * math.sqrt(3) / 2.0

    min_x = min(xs) - half_w - padding
    max_x = max(xs) + half_w + padding
    min_y = min(ys) - half_h - padding
    max_y = max(ys) + half_h + padding
    W = max_x - min_x
    H = max_y - min_y
    ox, oy = -min_x, -min_y   # offset to shift everything into positive coords

    bot_hexes: dict[int, object] = {bot.pid: bot.position for bot in state.bots.values()}
    occupied: set = set(bot_hexes.values())

    parts: list[str] = [
        f'<svg xmlns="http://www.w3.org/2000/svg" '
        f'viewBox="0 0 {W:.1f} {H:.1f}" width="100%" '
        f'style="display:block;background:{config.CANVAS_BG}">',
    ]

    # ── 1. Hex tiles ──────────────────────────────────────────────────────
    for h in state.grid:
        raw_cx, raw_cy = centers[h]
        cx, cy = raw_cx + ox, raw_cy + oy

        paint_pid = state.tile_pids.get(h, 0)
        is_occupied = h in occupied

        # Tiles the bot is standing on get a brighter accent fill
        if is_occupied:
            pid = next(p for p, pos in bot_hexes.items() if pos == h)
            fill   = _BOT_BRIGHT[pid]
            stroke = _TILE_STROKE[paint_pid]
            sw     = 2.0
        else:
            fill   = _TILE_FILL[paint_pid]
            stroke = _TILE_STROKE[paint_pid]
            sw     = 1.2

        corners = flat_hex_corners(cx, cy, hex_size - 0.8)
        pts = " ".join(f"{x:.2f},{y:.2f}" for x, y in corners)
        parts.append(
            f'<polygon points="{pts}" fill="{fill}" '
            f'stroke="{stroke}" stroke-width="{sw}"/>'
        )

    # ── 2. Bot markers ────────────────────────────────────────────────────
    for bot in state.bots.values():
        if bot.position not in centers:
            continue
        raw_cx, raw_cy = centers[bot.position]
        cx, cy = raw_cx + ox, raw_cy + oy
        r_outer = hex_size * 0.40
        r_inner = hex_size * 0.20
        bot_color = _BOT_FILL[bot.pid]

        # Outer ring (white halo)
        parts.append(
            f'<circle cx="{cx:.2f}" cy="{cy:.2f}" r="{r_outer + 2.5:.2f}" '
            f'fill="none" stroke="white" stroke-width="2" opacity="0.6"/>'
        )
        # Body
        parts.append(
            f'<circle cx="{cx:.2f}" cy="{cy:.2f}" r="{r_outer:.2f}" '
            f'fill="{bot_color}"/>'
        )
        # Inner dot for depth
        parts.append(
            f'<circle cx="{cx:.2f}" cy="{cy:.2f}" r="{r_inner:.2f}" '
            f'fill="rgba(0,0,0,0.3)"/>'
        )
        # Player number
        fs = max(9, int(hex_size * 0.36))
        parts.append(
            f'<text x="{cx:.2f}" y="{cy:.2f}" '
            f'text-anchor="middle" dominant-baseline="central" '
            f'fill="white" font-size="{fs}" font-weight="700" '
            f'font-family="monospace" opacity="0.9">{bot.pid}</text>'
        )

    parts.append("</svg>")
    return "\n".join(parts)

File: js/renderer/hex-renderer.js
/**
 * Converts a GameState snapshot to an SVG string.
 * Port of frontend/hex_renderer.py.
 */
import { HEX_DIRECTIONS, axialToPixel, hexCorners } from '../engine/hex-grid.js';
import config from '../config.js';

function facingAngleRad(facing) {
  const d = HEX_DIRECTIONS[((facing % 6) + 6) % 6];
  const [vx, vy] = axialToPixel(d.q, d.r, 1.0);
  return Math.atan2(vy, vx);
}

function arrowTrianglePoints(cx, cy, a0, hexSize) {
  const fx = Math.cos(a0), fy = Math.sin(a0);
  const px = -fy, py = fx;
  const baseBack = hexSize * 0.55;
  const tipForward = hexSize * 0.6;
  const halfBase = hexSize * 0.5;
  return [
    [cx + tipForward * fx, cy + tipForward * fy],
    [cx - baseBack * fx + halfBase * px, cy - baseBack * fy + halfBase * py],
    [cx - baseBack * fx - halfBase * px, cy - baseBack * fy - halfBase * py],
  ];
}

export function renderHexGrid(state, hexSize) {
  hexSize = hexSize ?? 26;

  const tileFill = {
    0: config.TILE_NONE_COLOR,
    1: config.PLAYER_TILE_COLORS[1],
    2: config.PLAYER_TILE_COLORS[2],
  };
  const tileStroke = { 0: config.TILE_STROKE_COLOR, 1: '#8b2e06', 2: '#065066' };
  const botFill = config.PLAYER_BOT_COLORS;
  const botBright = config.PLAYER_BRIGHT_COLORS;
  const padding = hexSize * 1.8;

  const centers = new Map();
  for (const h of state.grid.values()) {
    centers.set(h.key, axialToPixel(h.q, h.r, hexSize));
  }

  let minX = Infinity, maxX = -Infinity, minY = Infinity, maxY = -Infinity;
  for (const [cx, cy] of centers.values()) {
    if (cx < minX) minX = cx;
    if (cx > maxX) maxX = cx;
    if (cy < minY) minY = cy;
    if (cy > maxY) maxY = cy;
  }
  const halfW = hexSize * Math.sqrt(3) / 2;
  const halfH = hexSize;
  minX -= halfW + padding;
  maxX += halfW + padding;
  minY -= halfH + padding;
  maxY += halfH + padding;
  const W = maxX - minX;
  const H = maxY - minY;
  const ox = -minX, oy = -minY;

  const botHexes = new Map();
  const occupied = new Set();
  for (const bot of state.bots.values()) {
    botHexes.set(bot.pid, bot.position.key);
    occupied.add(bot.position.key);
  }

  const parts = [
    `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 ${W.toFixed(1)} ${H.toFixed(1)}" width="100%" style="display:block;background:${config.CANVAS_BG}">`,
  ];

  // Hex tiles
  for (const h of state.grid.values()) {
    const [rawCx, rawCy] = centers.get(h.key);
    const cx = rawCx + ox, cy = rawCy + oy;
    const paintPid = state.tilePids.get(h.key) || 0;
    const isOccupied = occupied.has(h.key);

    let fill, stroke, sw;
    if (isOccupied) {
      let pid = 0;
      for (const [p, posKey] of botHexes) {
        if (posKey === h.key) { pid = p; break; }
      }
      fill = botBright[pid];
      stroke = tileStroke[paintPid];
      sw = 2.0;
    } else {
      fill = tileFill[paintPid];
      stroke = tileStroke[paintPid];
      sw = 1.2;
    }

    const corners = hexCorners(cx, cy, hexSize - 0.8);
    const pts = corners.map(([x, y]) => `${x.toFixed(2)},${y.toFixed(2)}`).join(' ');
    parts.push(`<polygon points="${pts}" fill="${fill}" stroke="${stroke}" stroke-width="${sw}"/>`);
  }

  // Bot markers
  const display = config.BOT_DISPLAY_TYPE;
  for (const bot of state.bots.values()) {
    if (!centers.has(bot.position.key)) continue;
    const [rawCx, rawCy] = centers.get(bot.position.key);
    const cx = rawCx + ox, cy = rawCy + oy;
    const botColor = botFill[bot.pid];

    if (display === 'triangles') {
      const a0 = facingAngleRad(bot.facing);
      const triPts = arrowTrianglePoints(cx, cy, a0, hexSize);
      const ptsStr = triPts.map(([x, y]) => `${x.toFixed(2)},${y.toFixed(2)}`).join(' ');
      parts.push(
        `<polygon points="${ptsStr}" fill="none" stroke="white" stroke-width="2.5" opacity="0.6" stroke-linejoin="round"/>`,
      );
      parts.push(
        `<polygon points="${ptsStr}" fill="${botColor}" stroke="rgba(0,0,0,0.35)" stroke-width="1" stroke-linejoin="round"/>`,
      );
    } else {
      const rOuter = hexSize * 0.40;
      const rInner = hexSize * 0.20;
      parts.push(`<circle cx="${cx.toFixed(2)}" cy="${cy.toFixed(2)}" r="${(rOuter + 2.5).toFixed(2)}" fill="none" stroke="white" stroke-width="2" opacity="0.6"/>`);
      parts.push(`<circle cx="${cx.toFixed(2)}" cy="${cy.toFixed(2)}" r="${rOuter.toFixed(2)}" fill="${botColor}"/>`);
      parts.push(`<circle cx="${cx.toFixed(2)}" cy="${cy.toFixed(2)}" r="${rInner.toFixed(2)}" fill="rgba(0,0,0,0.3)"/>`);
    }

    const fs = Math.max(9, Math.floor(hexSize * 0.36));
    parts.push(
      `<text x="${cx.toFixed(2)}" y="${cy.toFixed(2)}" text-anchor="middle" dominant-baseline="central" fill="white" font-size="${fs}" font-weight="700" font-family="monospace" opacity="0.9">${bot.pid}</text>`,
    );
  }

  parts.push('</svg>');
  return parts.join('\n');
}
